### ISSUE TITLE
Can I mark String[] as source?

### ISSUE THREAD

**Issue by njupt-moon** [secure-software-engineering/soot-infoflow-android#114]
@StevenArzt 
Hi,
just as the title mentioned, can I mark String array as source?
Such as, 
$r1 = newarray (java.lang.String)[2];
$r1[0] = "android.permission.WRITE_EXTERNAL_STORAGE";
$r1[1] = "android.permission.READ_PHONE_STATE";
here I want to treat newarray (java.lang.String)[2] as source.
Thanks!

**Comment by StevenArzt** [comment]
FlowDroid Abstracts from the definition if what sources and sinks are through an interface called ISourceSinkManager. The default implementation of this interface cannot treat simple array creations as sources. You are, however, free to create your own implementation of ISourceSinkManager that adds this feature. The easiest way would be to derive an own class from the existing implementation so that you can just extend instead of having to re-implement what is already there.

**Comment by njupt-moon** [comment]
@StevenArzt 
Thanks, there is one more question, how can I deal with the constant value within an array? Such as "android.permission.WRITE_EXTERNAL_STORAGE" & "android.permission.READ_PHONE_STATE" in array $r1, can I recognize the constant value within an array? I have tried ValueBox.getValue, but it fails.

**Comment by StevenArzt** [comment]
If you have the assignment with the array reference on the left side and the constant string on the right side, you should be able to cast the right side to StringConstant and read the value from there.

**Comment by njupt-moon** [comment]
@StevenArzt 
Thanks!
While I am dealing with android application with soot, I want to treat some classes in apk as library classes, I have tried such code snapshot:
args = new String[] {"-android-jars", sdkPlatforms, "-process-dir", apkPath, "non-application", "-exclude", "android.support.v4"};
Options.v().parse(args);
Options.v().set_src_prec(Options.src_prec_apk);
Options.v().set_output_format(Options.output_format_none);
Options.v().set_allow_phantom_refs(true);
Options.v().set_whole_program(true);
Here I want to treat the class whose signature starts with "android.support.v4" as a library class, not an application class, but such code does not work. I have used Scene.v().getApplicationClasses(),  to check the answer, the class function <android.support.v4.content.ContextCompat: int checkSelfPermission(android.content.Context,java.lang.String)> still appears. How can I solve this problems? Thanks!

**Comment by njupt-moon** [comment]
@StevenArzt 
Hi!
As I want to specify some classes as library class, I have tried to modify SetupApplication.initializeSoot() function. Here is the code snapshot:
//
Options.v().set_app(true);
List<String> exclude = new ArrayList<String>();
exclude.add("android.app.Activity.");
//
Options.v().set_no_bodies_for_excluded(true);
Options.v().set_allow_phantom_refs(true);
Options.v().set_output_format(Options.output_format_none);
Options.v().set_whole_program(true);
Options.v().set_process_dir(Collections.singletonList(apkFileLocation));
Options.v().set_soot_classpath(getClasspath());
if (forceAndroidJar)
	Options.v().set_force_android_jar(androidJar);
else
	Options.v().set_android_jars(androidJar);
Options.v().set_src_prec(Options.src_prec_apk_class_jimple);
Options.v().set_keep_line_number(false);
Options.v().set_keep_offset(false);
Main.v().autoSetOptions(); 
I just add the first five lines to the original function, but it seems that these codes do not work at all, how can I solve this problem? Thanks!

**Comment by StevenArzt** [comment]
You have conflicting arguments. Set_app(true) puts Soot into application mode, i.e., treats all classes as application classes. This overwrites your excludes. Remove the application class line and you should be fine.

**Comment by njupt-moon** [comment]
@StevenArzt 
Thanks! I will try it. I still have a question.
Here is an edge in a callgraph.
VIRTUAL edge: virtualinvoke $r0.<com.google.android.apps.books.app.UploadsActivityAlias: void requestPermissions(java.lang.String[],int)>($r1, 0) in <com.google.android.apps.books.app.UploadsActivityAlias: void askForStoragePermission()> ==> <android.app.Activity: void requestPermissions(java.lang.String[],int)>
The following is the body of the method askForStoragePermission().
<com.google.android.apps.books.app.UploadsActivityAlias: void askForStoragePermission()>
    private void askForStoragePermission()
    {
        com.google.android.apps.books.app.UploadsActivityAlias $r0;
        java.lang.String[] $r1;
        $r0 := @this: com.google.android.apps.books.app.UploadsActivityAlias;
        $r1 = newarray (java.lang.String)[1];
        $r1[0] = "android.permission.READ_EXTERNAL_STORAGE";
        virtualinvoke $r0.<com.google.android.apps.books.app.UploadsActivityAlias: void requestPermissions(java.lang.String[],int)>($r1, 0);
        return;
    }
I have already marked the method <android.app.Activity: void requestPermissions(java.lang.String[],int)> as sink, and I have marked the ArrayNewExpr as source.
Actually, method in invoke statement virtualinvoke $r0.<com.google.android.apps.books.app.UploadsActivityAlias: void requestPermissions(java.lang.String[],int)>($r1, 0); is inherented from the sink method <android.app.Activity: void requestPermissions(java.lang.String[],int)>, but why flowdroid failed to find this sink method?
